src/community/jdbcstore/src/main/java/org/geoserver/jdbcstore/internal/JDBCDirectoryStructure.java

```diff
@@ -0,0 +1,406 @@
+        }
+        
+        
+        public InputStream getContent() {
+            InputStream is = helper.blobQuery(TABLE_RESOURCE, new PathSelector(path), CONTENT);
+            if (is == null) {
+                throw new IllegalStateException("Could not find content for entry " + toString());
+            }
+            return is;
+        }
+        
+        public void setContent(InputStream is) {
+            if (helper.updateQuery(TABLE_RESOURCE, new PathSelector(path), 
+                    new Assignment<InputStream>(CONTENT, is),
+                    new Assignment<Timestamp>(LAST_MODIFIED, 
+                          new Timestamp(new java.util.Date().getTime()))) <= 0) {
+                LOGGER.warning("Unable to write content to entry " + toString());
+            }
+        }
+        
+        public void createDirectory() {
+            int parentOid = 0;
+            for (String name : path) {
+                Map<String, Object> record = helper.selectQuery(TABLE_RESOURCE, 
+                        new ChildSelector(parentOid, name), OID, DIRECTORY);
+                
+                if (record == null) {
+                    parentOid = helper.insertQuery(TABLE_RESOURCE, 
+                            new Assignment<String>(NAME, name), 
+                            new Assignment<Integer>(PARENT, parentOid));                    
+                } else {
+                    if (!(Boolean) record.get(DIRECTORY.getFieldName())) {
+                        throw new IllegalStateException(
+                                "Could not create directory at " + toString() + 
+                                ": one of its parents exists and is not a directory.");
+                    }
+
+                    parentOid = (Integer) record.get(OID.getFieldName());
+                }
+            }
+        }
+        
+        public boolean createResource() {
+            Boolean dir = isDirectory();
+            
+            if (dir != null) {
+                if (dir) {
+                    throw new IllegalStateException("Could not create resource at " + toString() + 
+                            ": already a directory.");
+                } else {
+                    return false;
+                }
+            }
+            
+            Entry parent = getParent();
+            try {
+                parent.createDirectory();         
+            } catch (IllegalStateException e) {
+                throw new IllegalStateException("Could not create resource at " + toString() + 
+                        ": could not create parent directory.", e);
+            }
+            
+            ByteArrayInputStream is = new ByteArrayInputStream(new byte[0]);
+            Integer oid = helper.insertQuery(TABLE_RESOURCE,
+                    new Assignment<String>(NAME, getName()), 
+                    new Assignment<Integer>(PARENT, parent.getOid()), 
+                    new Assignment<InputStream>(CONTENT, is));
+            try {
```

Comment:
This needs some work. First try with resources, and the oid null test only applies if insertQuery fails; does it throw an exception? If so, then catch it and rethrow an IllegalStateException. Does that make more sense? And there is a bunch of effort around is.close() that could be handled implicilty, not to mention the fact that is.close() will never, ever fail; from the javadoc: "Closing a ByteArrayInputStream has no effect.".